Add new e2e test merchant import products via CSV file

veljkho opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #28489

### How to test the changes in this Pull Request:

npx wc-e2e test:e2e ./tests/e2e/specs/wp-admin/test-product-import-csv.js

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

veljkho committed:
First part of the test

veljkho committed:
Add first test import products

veljkho committed:
Add new scenario and update changelogs

veljkho committed:
Merge branch 'trunk' into e2e/e2e-merchant-import-products

veljkho committed:
Fix spaces in the code

veljkho committed:
Improve test scenario

veljkho committed:
Fix test and correct scenario

veljkho committed:
Add after all scenario

veljkho committed:
Add removing products before all

rrennick commented:
@veljkho There is a function in the local jest setup script to delete all products https://github.com/woocommerce/woocommerce/blob/trunk/tests/e2e/config/jest.setup.js#L46. Can you move that to the utilities package and use it to delete all the products in this test as well?

veljkho committed:
Remove removing products from before all

veljkho commented:
> @veljkho There is a function in the local jest setup script to delete all products https://github.com/woocommerce/woocommerce/blob/trunk/tests/e2e/config/jest.setup.js#L46. Can you move that to the utilities package and use it to delete all the products in this test as well?

Hey @rrennick 
I'm not sure we need that, because it should be expected to have clean data before running every test but I realized that there were products left after my test and that they could conflict with other tests, so just to make sure they were deleted I added this `afterAll` scenario. My question is, can we make `remove all data before every test` really working as expected so I don't need to put `afterAll`?

rrennick commented:
> My question is, can we make remove all data before every test really working as expected so I don't need to put afterAll?

We should use `beforeAll`. Other projects using the tests might not run them in the same order or could intersperse our tests with theirs. The tests will be more reliable for everyone if they reset to a known state at the beginning of the test.

rrennick commented:
The `beforeAll` in the jest setup file runs before each test file vs before each `describe`. In the default run we only have 4 tests files. For some of the tests like this one we will need to clear the products before that test.

veljkho committed:
Remove after all scenario

veljkho commented:
I just removed `afterAll` scenario and added the existing `moveAllItemsToTrash` which seemed a good option to me, please review @rrennick. This way we keep less code.

tammullen commented:
I believe anything we add to `sample-data` will be included in the core plugin? If it is included it would therefore increase the file size of the plugin. So I think we should try find a different location for `sample_products_override.csv` within `tests/e2e`.

veljkho commented:
@tammullen If it makes a difference, I can create a new folder in `tests/e2e` called `sample-data` with the file `sample_products_override.csv`? The file size is 18.7 KB, but I agree to move.

tammullen commented:
> new folder in tests/e2e called sample-data with the file sample_products_override.csv

As there is references to the data contained in `sample_products_override.csv` in `core-tests/specs/merchant/wp-admin-product-import-csv.test.js` I think it would be better having that folder within `core-tests` so that it is included in that package?

veljkho committed:
Created test data folder and moved csv file

veljkho commented:
Folder `test-data` created in `core-tests` and test updated.

rrennick commented on tests/e2e/core-tests/specs/merchant/wp-admin-product-import-csv.test.js:
```diff
@@ -0,0 +1,119 @@
+} = require( '@jest/globals' );
+
+const filePath = '../../../sample-data/sample_products.csv';
+const filePathOverride = '../../../tests/e2e/core-tests/test-data/sample_products_override.csv';
```
This path won't work in the package. When another project uses the package it is installed in `node_modules/@woocommerce/e2e-core-tests`. They won't have a `core-tests` folder. https://github.com/woocommerce/woocommerce/blob/trunk/tests/e2e/env/utils/app-root.js has the logic you can use to determine which folder path to use. You could make a `getCoreTestRoot()` in `core-tests` which might be re-used in the future.

veljkho replied:
Thanks, @rrennick I did that way!

rrennick commented:
@veljkho Nice work on this. It tested great for me locally. I only found the one item that needs to be addressed.

> I just removed afterAll scenario and added the existing moveAllItemsToTrash which seemed a good option to me

That sounds good. We can add the utility functions in the PR for #28349.

veljkho committed:
Added core tests root and updated test

rrennick commented on tests/e2e/core-tests/core-tests-root.js:
```diff
@@ -0,0 +1,22 @@
+     // Typically will be in node_modules/, but WooCommerce
+     // uses a local file path (tests/e2e/env).
+     let coreTestsPath = false;
+     const moduleDir = path.dirname( require.resolve( '@woocommerce/e2e-environment' ) );
```
This should be `@woocommerce/e2e-core-tests` because we are sure that package exists. We still need to eliminate the `tests/e2e/core-tests` string below. I think it will work if this just returns `moduleDir`.

rrennick commented on tests/e2e/core-tests/specs/merchant/wp-admin-product-import-csv.test.js:
```diff
@@ -0,0 +1,122 @@
+
+const path = require( 'path' );
+const coreTestsPath = getCoreTestsRoot();
+const filePath = path.resolve( coreTestsPath, 'sample-data/sample_products.csv' );
```
With previous comment:
```
const filePath = path.resolve( coreTestsPath.appRoot, 'sample-data/sample_products.csv' );
const filePathOverride = path.resolve( coreTestsPath.packageRoot, '/test-data/sample_products_override.csv' );
```

I did not test that. It may need some adjustments.

veljkho replied:
I did and it worked 👍 @rrennick

rrennick replied:
@veljkho Looks good. This worked great for me locally.

veljkho committed:
Updated core-tests-root file

zhongruige commented on tests/e2e/core-tests/test-data/sample_products_override.csv:
```diff
@@ -0,0 +1,26 @@
+﻿ID,Type,SKU,Name,Published,"Is featured?","Visibility in catalog","Short description",Description,"Date sale price starts","Date sale price ends","Tax status","Tax class","In stock?",Stock,"Backorders allowed?","Sold individually?","Weight (lbs)","Length (in)","Width (in)","Height (in)","Allow customer reviews?","Purchase note","Sale price","Regular price",Categories,Tags,"Shipping class",Images ...
```
The product image URLs (such as `https://woocommercecore.mystagingwebsite.com/wp-content/uploads/2017/12/logo-1.jpg` should be fairly stable, right? I believe the importer fails if the images can't be found, so just wanted to check to verify we expect this site to be around for a while (considering those images were uploaded in 2017 I assume so, but figured I'd double check).

rrennick replied:
> The product image URLs (such as https://woocommercecore.mystagingwebsite.com/wp-content/uploads/2017/12/logo-1.jpg) should be fairly stable, right?

Those are the URLs that are distributed with WooCommerce: https://github.com/woocommerce/woocommerce/blob/trunk/sample-data/sample_products.csv

zhongruige replied:
Ah! Thank you for clarifying that @rrennick!

veljkho committed:
Merge branch 'trunk' into e2e/e2e-merchant-import-products

veljkho committed:
Merge branch 'trunk' into e2e/e2e-merchant-import-products

veljkho committed:
Merge branch 'trunk' into e2e/e2e-merchant-import-products

tammullen reviewed (approved):
Thanks for this! Runs great

```
 PASS  ../specs/wp-admin/test-product-import-csv.js (57.655s)
  Import Products from a CSV file
    ✓ can upload the CSV file and import products (25311ms)
    ✓ can override the existing products via CSV import (18009ms)

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        57.754s
```
